main.go

```diff
@@ -3,160 +3,257 @@ package main
 	}
 }
 
-func init() {
+func initConfig() {
 	viper.SetEnvPrefix("audit")
-	homedir, err := homedir.Dir()
-	if err != nil {
-		logger.Fatal(err)
-	}
-	rootCmd.PersistentFlags().StringP("kubecfg", "k", homedir+"/.kube/config", "kubecfg path to the cluster to account")
-	rootCmd.PersistentFlags().Duration("fetch-interval", 10*time.Second, "interval for checking availability of target")
+	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
 	viper.AutomaticEnv()
-	err = viper.BindPFlags(rootCmd.PersistentFlags())
-	if err != nil {
-		logger.Fatal(err)
+
+	viper.SetConfigType(cfgFileType)
+
+	if cfgFile != "" {
+		viper.SetConfigFile(cfgFile)
+		if err := viper.ReadInConfig(); err != nil {
+			logger.Errorw("Config file path set explicitly, but unreadable", "error", err)
+			os.Exit(1)
+		}
+	} else {
+		viper.SetConfigName("config")
+		viper.AddConfigPath("/etc/" + moduleName)
+		viper.AddConfigPath("$HOME/." + moduleName)
+		viper.AddConfigPath(".")
+		if err := viper.ReadInConfig(); err != nil {
+			usedCfg := viper.ConfigFileUsed()
+			if usedCfg != "" {
+				logger.Errorw("Config file unreadable", "config-file", usedCfg, "error", err)
+				os.Exit(1)
+			}
+		}
+	}
+
+	usedCfg := viper.ConfigFileUsed()
+	if usedCfg != "" {
+		logger.Infow("Read config file", "config-file", usedCfg)
 	}
 }
 
-func run() {
-	client, err := loadClient(viper.GetString("kubecfg"))
-	if err != nil {
-		logger.Fatalw("Unable to connect to k8s", "Error", err)
+func initLogging() {
+	level := zap.InfoLevel
+
+	if viper.IsSet("log-level") {
+		err := level.UnmarshalText([]byte(viper.GetString("log-level")))
+		if err != nil {
+			log.Fatalf("can't initialize zap logger: %v", err)
+		}
 	}
 
-	fetchInterval := viper.GetDuration("fetch-interval")
+	cfg := zap.NewProductionConfig()
+	cfg.Level = zap.NewAtomicLevelAt(level)
 
-	// Apparently we now need to pass a context co k8s client API
-	kubectx, kubecancel := context.WithCancel(context.Background())
-	defer kubecancel()
+	log.Printf("Log level: %s", cfg.Level)
 
-	// initialising variables used in pod loop
-	oldPodIP := ""
-	var oldPodDate time.Time
-	ctx, cancel := context.WithCancel(context.Background())
-	forwarderKilled := make(chan struct{})
+	l, err := cfg.Build()
+	if err != nil {
+		log.Fatalf("can't initialize zap logger: %v", err)
+	}
 
-	logger.Infow("Initial old values", "oldPodIP", oldPodIP, "oldPodDate", oldPodDate)
+	logger = l.Sugar()
+}
+
+func initSignalHandlers() {
+	stop = signals.SetupSignalHandler()
+}
 
-	labelMap := map[string]string{"app": podname}
-	opts := metav1.ListOptions{
-		LabelSelector: labels.SelectorFromSet(labelMap).String(),
+func run(opts *Opts) error {
+	logger.Infow("Options", "opts", opts)
+	// initialise our synchronisation channels
+	forwarderKilledChan = make(chan struct{})
+	killForwarderChan = make(chan struct{}, 1)
+
+	// Prepare K8s
+	client, err := loadClient(opts.KubeCfg)
+	if err != nil {
+		logger.Fatalw("Unable to connect to k8s", "Error", err)
```

Comment:
After the entire init phase there should be no need to call `Fatal` anymore, just return the error and let the underlying layer handle the issue.

Reply:
Changed - not sure if this is what you had in mind?

Reply:
Now doing return(err)